doc/users-guide/datapar/promotion.rst

```diff
@@ -0,0 +1,324 @@
+which takes a ``real`` value by reference and negates it:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :caption:
+  :language: chapel
+  :lines: 1-3
+
+In addition to calling the function with a single ``real`` variable,
+we can also call the function in a promoted manner using an array of
+reals, as follows:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 9-10
+
+If we print out the array after making this call, we can see that each
+element has been negated, as expected:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.good
+   :language: text
+   :lines: 3
+
+Moreover, the execution will be equivalent to the following
+data-parallel forall-loop:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 14-15
+
+Nothing about the above requires using a 1D array.  We could also
+promote ``negate()`` by passing it a multidimensional array of reals:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 19-22
+
+which would again be equivalent to the loop:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 25-26
+
+
+Multi-Argument Promotion
+------------------------
+
+When multiple function arguments are promoted in a single call, the
+result is equivalent to a zippered forall-loop.  As an example,
+consider the following function which takes two ``real`` arguments by
+``ref`` and sorts them relative to one another using a swap
+assignment:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 30-33
+
+This function can be promoted using two arrays of reals as follows:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 35-38
+
+This promotion is equivalent to the following zippered forall-loop:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 45-46
+
+Either form will leave the arrays as follows:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.good
+   :language: text
+   :lines: 15-16
+
+Promoting multiple arguments simultaneously requires that the
+arguments can legally be zippered together, as in an explicit
+forall-loop.
+
+All of a function's arguments need not be promoted simultaneously.
+Any arguments that are promoted define the implicit, potentially
+zippered forall-loop.  All others are simply passed through to their
+formal argument normally.  For example, consider the following
+function which assigns its *y* argument into its *x* argument only if
+the third argument, *b*, is true:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 51-54
+
+The following four calls demonstrate that various combinations of
+arguments may be promoted, where *Mask* is an array of ``bool``
+values:
```

Comment:
Yes, but it would also be fine with me to not even bother to have 'Mask' as a way of reducing the number of examples here.